### PR TITLE
[ARCTIC-741]Improve the performance of iceberg table optimizing

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

Improve the performance of iceberg table optimizing.

## Brief change log

  - *Compact delete files in minor optimize processing.*
  - *Add an IcebergCombainedReader to speed up the reading process for optimizing.*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergMinorOptimizePlan.java [hunk 1]
```diff
@@ -26,12 +26,19 @@
 import com.netease.arctic.table.TableProperties;
 import org.apache.commons.collections.CollectionUtils;
 import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
 import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
```
Comment (wangtaohz) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergMinorOptimizePlan.java:33]: useless import

File: ams/ams-server/src/main/resources/sql/mysql/upgrade-0.3.2-to-0.4.0.sql [hunk 2]
```diff
@@ -15,11 +15,7 @@ ALTER TABLE file_info_cache ADD COLUMN `add_snapshot_sequence` bigint(20) NOT NU
 after `delete_snapshot_id`;
 ALTER TABLE snapshot_info_cache ADD COLUMN `snapshot_sequence` bigint(20) NOT NULL DEFAULT -1 COMMENT 'snapshot sequence' after `snapshot_id`;
 ALTER TABLE file_info_cache DROP COLUMN `watermark`;
-ALTER TABLE `optimize_task` ADD COLUMN `eq_delete_files` int(11) DEFAULT NULL COMMENT 'Eq-Delete file cnt';
-ALTER TABLE `optimize_task` ADD COLUMN `eq_delete_file_size` bigint(20) DEFAULT NULL COMMENT 'Eq-Delete file size in bytes';
-ALTER TABLE `optimize_history` ADD COLUMN `eq_delete_file_cnt_before` int(11) NOT NULL COMMENT 'Eq-Delete file cnt before optimizing';
-ALTER TABLE `optimize_history` ADD COLUMN `eq_delete_file_size_before` bigint(20) NOT NULL COMMENT 'Eq-Delete file size in bytes before optimizing';
-ALTER TABLE `optimize_file` CHANGE `file_type` `content_type` varchar(32) NOT NULL COMMENT 'File type: BASE_FILE, INSERT_FILE, EQ_DELETE_FILE, POS_DELETE_FILE, FILE_SCAN_TASK';
+ALTER TABLE `optimize_file` CHANGE `file_type` `content_type` varchar(32) NOT NULL COMMENT 'File type: BASE_FILE, INSERT_FILE, EQ_DELETE_FILE, POS_DELETE_FILE';
```
[the remark this answers is not in the document]
Comment (hzluting) [reply ams/ams-server/src/main/resources/sql/mysql/upgrade-0.3.2-to-0.4.0.sql:18]: > The column definition seems the same with version 0.3.2

No, I have changed this column in #581 for version 0.4.
Now, the comment has changed.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergMinorOptimizePlan.java [hunk 3]
```diff
@@ -93,21 +147,83 @@ protected List<BaseOptimizeTask> collectTask(String partition) {
...
+    }
+
+    List<FileScanTask> allNeedOptimizeTask = new ArrayList<>();
+    for (DeleteFile needOptimizeDeleteFile : needOptimizeDeleteFiles) {
+      allNeedOptimizeTask.addAll(deleteDataFileMap.get(needOptimizeDeleteFile.path().toString()));
+    }
+
+    allNeedOptimizeTask = filterRepeatFileScanTask(allNeedOptimizeTask);
+
+    List<FileScanTask> sortedNeedOptimizeTask = allNeedOptimizeTask.stream()
```
[the remark this answers is not in the document]
Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergMinorOptimizePlan.java:204]: Ok, I will remove it.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/utils/SequenceNumberFetcher.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import com.netease.arctic.scan.TableEntriesScan;
+import org.apache.iceberg.FileContent;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+import java.util.Map;
+
+public class SequenceNumberFetcher {
```
[the remark this answers is not in the document]
Comment (wangtaohz) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/utils/SequenceNumberFetcher.java:30]: Ok, I will

**Review by wangtaohz — APPROVED** [review]

**Pull request merged**